autogpt_platform/frontend/src/hooks/useAdminImpersonation.ts

```diff
@@ -0,0 +1,83 @@
+"use client";
+
+import { useState, useCallback } from "react";
+import { environment } from "@/services/environment";
+import { IMPERSONATION_STORAGE_KEY } from "@/lib/constants";
+
+interface AdminImpersonationState {
+  isImpersonating: boolean;
+  impersonatedUserId: string | null;
+}
+
+interface AdminImpersonationActions {
+  startImpersonating: (userId: string) => void;
+  stopImpersonating: () => void;
+}
+
+type AdminImpersonationHook = AdminImpersonationState &
+  AdminImpersonationActions;
+
+/**
+ * Get the initial impersonation state from sessionStorage synchronously to prevent UI flicker.
+ */
+function getInitialImpersonationState(): string | null {
+  if (!environment.isClientSide()) {
+    return null;
+  }
+
+  try {
+    return sessionStorage.getItem(IMPERSONATION_STORAGE_KEY);
+  } catch (error) {
+    console.error("Failed to read initial impersonation state:", error);
+    return null;
+  }
+}
+
+/**
+ * Hook for managing admin user impersonation state.
+ *
+ * Provides functions to start/stop impersonating users and automatically
+ * sets the X-Act-As-User-Id header for API requests via sessionStorage.
+ * Triggers a full page refresh to ensure all data is updated correctly.
+ */
+export function useAdminImpersonation(): AdminImpersonationHook {
```

Comment:
If this is, for now, it's only used on `<AdminImpersionationPanel />`, let's move it there. We can move it to the shared hooks once it is used on other places too.